pkg/artifactory/resource/webhook/resource_artifactory_custom_webhook.go

```diff
@@ -0,0 +1,454 @@
+						Elem: &schema.Schema{
+							Type:             schema.TypeString,
+							ValidateDiagFunc: validation.ToDiagFunc(validation.StringMatch(regexp.MustCompile("^[a-zA-Z_][a-zA-Z0-9_]*$"), "Secret name must match '^[a-zA-Z_][a-zA-Z0-9_]*$'\"")),
+						},
+						Description: "A set of sensitive values that will be injected in the request (headers and/or payload), comprise of key/value pair.",
+					},
+					"proxy": {
+						Type:     schema.TypeString,
+						Optional: true,
+						ValidateDiagFunc: validator.All(
+							validator.StringIsNotEmpty,
+							validator.StringIsNotURL,
+						),
+						Description: "Proxy key from Artifactory UI (Administration -> Proxies -> Configuration)",
+					},
+					"http_headers": {
+						Type:        schema.TypeMap,
+						Optional:    true,
+						Elem:        &schema.Schema{Type: schema.TypeString},
+						Description: "HTTP headers you wish to use to invoke the Webhook, comprise of key/value pair. Used in custom webhooks.",
+					},
+					"payload": {
+						Type:             schema.TypeString,
+						Optional:         true,
+						ValidateDiagFunc: validation.ToDiagFunc(validation.StringIsNotEmpty),
+						Description:      "This attribute is used to build the request body. Used in custom webhooks",
+					},
+				},
+			},
+		},
+	}
+}
+
+type CustomBaseParams struct {
+	Key         string          `json:"key"`
+	Description string          `json:"description"`
+	Enabled     bool            `json:"enabled"`
+	EventFilter EventFilter     `json:"event_filter"`
+	Handlers    []CustomHandler `json:"handlers"`
+}
+
+func (w CustomBaseParams) Id() string {
+	return w.Key
+}
+
+type CustomHandler struct {
+	HandlerType string         `json:"handler_type"`
+	Url         string         `json:"url"`
+	Secrets     []KeyValuePair `json:"secrets"`
+	Proxy       string         `json:"proxy"`
+	HttpHeaders []KeyValuePair `json:"http_headers"`
+	Payload     string         `json:"payload,omitempty"`
+}
+
+type SecretName struct {
+	Name string `json:"name"`
+}
+
+var packKeyValuePairNoValue = func(keyValuePairs []KeyValuePair, d *schema.ResourceData, url string) map[string]interface{} {
+	KVPairs := make(map[string]interface{})
+	// Get secrets from TF state
+	var secrets map[string]interface{}
+	if v, ok := d.GetOk("handler"); ok {
+		handlers := v.(*schema.Set).List()
+		for _, handler := range handlers {
+			h := handler.(map[string]interface{})
+			// if url match, merge secret maps
+			if h["url"] == url {
+				secrets = utilsdk.MergeMaps(secrets, h["secrets"].(map[string]interface{}))
+			}
+		}
+	}
+	// We assign secret the value from the state, because it's not returned in the API body response
+	for _, keyValuePair := range keyValuePairs {
+		if v, ok := secrets[keyValuePair.Name]; ok {
+			KVPairs[keyValuePair.Name] = v.(string)
+		}
+	}
+
+	return KVPairs
+}
+
+func ResourceArtifactoryCustomWebhook(webhookType string) *schema.Resource {
+
+	var domainCriteriaLookup = map[string]interface{}{
```

Comment:
These lookup maps can be moved out of the func and into the `webhook` package level so we don't have duplicates with regular webhook resource.